opm/autodiff/FlowMainEbos.hpp

```diff
@@ -86,21 +87,174 @@ namespace Opm
+
+            // Geological properties
+            Base::use_local_perm_ = true;
+            geoprops_.reset(new DerivedGeology(grid,
+                                               *Base::fluidprops_,
+                                               *Base::eclipse_state_,
+                                               Base::use_local_perm_,
+                                               Base::gravity_.data()));
+
+        }
+
+        // Initialise the reservoir state. Updated fluid props for SWATINIT.
+        // Writes to:
+        //   state_
+        //   threshold_pressures_
+        //   fluidprops_ (if SWATINIT is used)
+        void setupState()
+        {
+            const PhaseUsage pu = Opm::phaseUsageFromDeck(*deck_);
+            const Grid& grid = Base::grid_init_->grid();
+
+            // Need old-style fluid object for init purposes (only).
+            BlackoilPropertiesFromDeck props(*deck_,
+                                             *Base::eclipse_state_,
+                                             Base::material_law_manager_,
+                                             grid.size(/*codim=*/0),
+                                             grid.globalCell().data(),
+                                             grid.logicalCartesianSize().data(),
+                                             param_);
+
+
+            // Init state variables (saturation and pressure).
+            if (param_.has("init_saturation")) {
+                state_.reset(new ReservoirState(grid.size(/*codim=*/0),
+                                                grid.numFaces(),
+                                                props.numPhases()));
+
+                initStateBasic(grid.size(/*codim=*/0),
+                               grid.globalCell().data(),
+                               grid.logicalCartesianSize().data(),
+                               grid.numFaces(),
+                               Opm::UgGridHelpers::faceCells(grid),
+                               Opm::UgGridHelpers::beginFaceCentroids(grid),
+                               Opm::UgGridHelpers::beginCellCentroids(grid),
+                               Grid::dimension,
+                               props, param_, gravity_[2], *state_);
+
+                initBlackoilSurfvol(Opm::UgGridHelpers::numCells(grid), props, *state_);
+
+                enum { Oil = BlackoilPhases::Liquid, Gas = BlackoilPhases::Vapour };
+                if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                    const int numPhases = props.numPhases();
+                    const int numCells  = Opm::UgGridHelpers::numCells(grid);
+
+                    // Uglyness 1: The state is a templated type, here we however make explicit use BlackoilState.
+                    auto& gor = state_->getCellData( BlackoilState::GASOILRATIO );
+                    const auto& surface_vol = state_->getCellData( BlackoilState::SURFACEVOL );
+                    for (int c = 0; c < numCells; ++c) {
```

Comment:
`numCells` seems wrong here (even in a parallel run) as there is no guarantee that the number of cells of the grid is equal to the number of active cells of the eclipse grid representation.